### PR TITLE
Mrossides/generic social proof

### PR DESCRIPTION
**Pull request by rossides**
1) Update the type of the graph used in SocialProof from NodeMetadataLeftIndexedMultiSegmentBipartiteGraph to LeftIndexedMultiSegmentBipartiteGraph. 

2) Update the naming to make it more generic.

3) Update the interface to take IDMask and RecommendationType on a per SocialProofGenerator base.

### CODE REVIEW

File: graphjet-core/src/main/java/com/twitter/graphjet/algorithms/IDMask.java [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+
+package com.twitter.graphjet.algorithms;
+
+public interface IDMask {
```
Comment (guimingTang) [inline graphjet-core/src/main/java/com/twitter/graphjet/algorithms/IDMask.java]: Could you copy the original description of id masks here? Easier to tell the usage of this class especially when it only contains a simple function

File: graphjet-core/src/main/java/com/twitter/graphjet/algorithms/IdentityIDMask.java [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.twitter.graphjet.algorithms;
+
+public class IdentityIDMask implements IDMask {
+  public long restore(long node) {
```
Comment (guimingTang) [inline graphjet-core/src/main/java/com/twitter/graphjet/algorithms/IdentityIDMask.java]: + @Override

File: graphjet-core/src/main/java/com/twitter/graphjet/algorithms/counting/tweet/TopSecondDegreeByCountTweetMetadataRecsGenerator.java [hunk 3]
```diff
@@ -46,12 +46,13 @@ private static void addToSocialProof(
     int maxTweetSocialProofSize
   ) {
     SmallArrayBasedLongToDoubleMap[] socialProofsByType = nodeInfo.getSocialProofs();
+    TweetIDMask tweetIDMask = new TweetIDMask();
```
Comment (guimingTang) [inline graphjet-core/src/main/java/com/twitter/graphjet/algorithms/counting/tweet/TopSecondDegreeByCountTweetMetadataRecsGenerator.java]: We can move this to be a class field

File: graphjet-core/src/main/java/com/twitter/graphjet/algorithms/counting/tweet/TopSecondDegreeByCountTweetRecsGenerator.java [hunk 4]
```diff
@@ -69,11 +69,12 @@ private TopSecondDegreeByCountTweetRecsGenerator() {
     }
 
     List<RecommendationInfo> outputResults = Lists.newArrayListWithCapacity(topResults.size());
+    TweetIDMask tweetIDMask = new TweetIDMask();
```
Comment (guimingTang) [inline graphjet-core/src/main/java/com/twitter/graphjet/algorithms/counting/tweet/TopSecondDegreeByCountTweetRecsGenerator.java]: Same, better to change it to a class field

File: graphjet-core/src/main/java/com/twitter/graphjet/algorithms/socialproof/MomentSocialProofGenerator.java [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+ */
+
+
+package com.twitter.graphjet.algorithms.socialproof;
+
+import com.twitter.graphjet.algorithms.IdentityIDMask;
+import com.twitter.graphjet.algorithms.RecommendationType;
+import com.twitter.graphjet.bipartite.LeftIndexedMultiSegmentBipartiteGraph;
+
+public class MomentSocialProofGenerator extends SocialProofGenerator {
```
Comment (guimingTang) [inline graphjet-core/src/main/java/com/twitter/graphjet/algorithms/socialproof/MomentSocialProofGenerator.java]: hmm, don't see this being used anywhere other than test.

File: graphjet-core/src/main/java/com/twitter/graphjet/algorithms/socialproof/TweetSocialProofGenerator.java [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+ */
+
+
+package com.twitter.graphjet.algorithms.socialproof;
+
+import com.twitter.graphjet.algorithms.RecommendationType;
+import com.twitter.graphjet.algorithms.TweetIDMask;
+import com.twitter.graphjet.bipartite.LeftIndexedMultiSegmentBipartiteGraph;
+
+public class TweetSocialProofGenerator extends SocialProofGenerator {
```
Comment (guimingTang) [inline graphjet-core/src/main/java/com/twitter/graphjet/algorithms/socialproof/TweetSocialProofGenerator.java]: Seems there is no use case for this class either

File: graphjet-core/src/main/java/com/twitter/graphjet/algorithms/IdentityIDMask.java [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.twitter.graphjet.algorithms;
+
+public class IdentityIDMask implements IDMask {
```
Comment (guimingTang) [inline graphjet-core/src/main/java/com/twitter/graphjet/algorithms/IdentityIDMask.java]: Looks like it's only used for Moments. Rename to MomentsIDMask?

File: graphjet-core/src/main/java/com/twitter/graphjet/algorithms/IDMask.java [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+  /*
+   * Return the original value of the node, before additional information was masked into it.
+   */
+  long restore(long node);
```
Comment (guimingTang) [inline graphjet-core/src/main/java/com/twitter/graphjet/algorithms/IDMask.java]: Explicitly declare visibility of the function

File: graphjet-core/src/main/java/com/twitter/graphjet/algorithms/TweetIDMask.java [hunk 9]
```diff
@@ -66,7 +62,7 @@ public static long promotion(long tweet) {
    * @param node the tweet id with bitmask
    * @return tweet id without the bitmask
    */
-  public static long restore(long node) {
+  public long restore(long node) {
```
Comment (guimingTang) [inline graphjet-core/src/main/java/com/twitter/graphjet/algorithms/TweetIDMask.java]: @Override

File: graphjet-core/src/main/java/com/twitter/graphjet/algorithms/counting/tweet/TopSecondDegreeByCountTweetMetadataRecsGenerator.java (hunk 3 above)
Comment (jerryjiangabc) [inline graphjet-core/src/main/java/com/twitter/graphjet/algorithms/counting/tweet/TopSecondDegreeByCountTweetMetadataRecsGenerator.java]: this function is being called hundreds of thousand times a second. this additional object creation will be roughly 1M bytes per second. why not use static function? or at least this should be a class variable?

File: graphjet-core/src/main/java/com/twitter/graphjet/algorithms/counting/tweet/TopSecondDegreeByCountTweetRecsGenerator.java (hunk 4 above)
Comment (jerryjiangabc) [inline graphjet-core/src/main/java/com/twitter/graphjet/algorithms/counting/tweet/TopSecondDegreeByCountTweetRecsGenerator.java]: ditto, static or class variable.

File: graphjet-core/src/main/java/com/twitter/graphjet/algorithms/counting/tweet/TopSecondDegreeByCountTweetMetadataRecsGenerator.java (hunk 3 above)
Comment (rossides) [inline graphjet-core/src/main/java/com/twitter/graphjet/algorithms/counting/tweet/TopSecondDegreeByCountTweetMetadataRecsGenerator.java]: Fixed

File: graphjet-core/src/main/java/com/twitter/graphjet/algorithms/counting/tweet/TopSecondDegreeByCountTweetRecsGenerator.java (hunk 4 above)
Comment (rossides) [inline graphjet-core/src/main/java/com/twitter/graphjet/algorithms/counting/tweet/TopSecondDegreeByCountTweetRecsGenerator.java]: Fixed

File: graphjet-core/src/main/java/com/twitter/graphjet/algorithms/IDMask.java (hunk 1 above)
Comment (rossides) [inline graphjet-core/src/main/java/com/twitter/graphjet/algorithms/IDMask.java]: Done

File: graphjet-core/src/main/java/com/twitter/graphjet/algorithms/IdentityIDMask.java (hunk 7 above)
Comment (rossides) [inline graphjet-core/src/main/java/com/twitter/graphjet/algorithms/IdentityIDMask.java]: Right now it's only used for Moments, but I will keep it generic for any other entity that is not using masks to encode information into the IDs (for example, URLs, Users etc)

File: graphjet-core/src/main/java/com/twitter/graphjet/algorithms/IdentityIDMask.java (hunk 2 above)
Comment (rossides) [inline graphjet-core/src/main/java/com/twitter/graphjet/algorithms/IdentityIDMask.java]: Fixed

File: graphjet-core/src/main/java/com/twitter/graphjet/algorithms/TweetIDMask.java (hunk 9 above)
Comment (rossides) [inline graphjet-core/src/main/java/com/twitter/graphjet/algorithms/TweetIDMask.java]: Fixed

File: graphjet-core/src/main/java/com/twitter/graphjet/algorithms/socialproof/SocialProofGenerator.java [hunk 10]
```diff
@@ -0,0 +1,140 @@
...
+
+          // If the current id is in the set of inputIds, we find and store its social proof.
+          if (inputIds.contains(rightNode) && socialProofTypes.contains(edgeType)) {
+            if (!socialProofs.containsKey(rightNode)) {
+              socialProofs.put(rightNode, new Byte2ObjectArrayMap<>());
+              socialProofWeights.put(rightNode, 0);
+            }
+            Byte2ObjectMap<LongSet> socialProofMap = socialProofs.get(rightNode);
+
+            // We sum the weights of incoming leftNodes as the weight of a rightNode.
```
Comment (jerryjiangabc) [inline graphjet-core/src/main/java/com/twitter/graphjet/algorithms/socialproof/SocialProofGenerator.java]: the weight of a rightNode => the weight of the rightNode.

File: graphjet-core/src/main/java/com/twitter/graphjet/algorithms/socialproof/SocialProofGenerator.java [hunk 11]
```diff
@@ -0,0 +1,140 @@
...
+            socialProofWeights.put(
+              rightNode,
+              weight + socialProofWeights.get(rightNode)
+            );
+
+            // Get the user set variable by the engagement type.
+            if (!socialProofMap.containsKey(edgeType)) {
+              socialProofMap.put(edgeType, new LongArraySet());
+            }
+            LongSet incomingLeftNodes = socialProofMap.get(edgeType);
```
Comment (jerryjiangabc) [inline graphjet-core/src/main/java/com/twitter/graphjet/algorithms/socialproof/SocialProofGenerator.java]: I prefer the original variable name, connectingUsers

File: graphjet-core/src/main/java/com/twitter/graphjet/algorithms/socialproof/SocialProofGenerator.java [hunk 12]
```diff
@@ -0,0 +1,140 @@
...
+
+  /**
+   * Collect social proofs for a given {@link SocialProofRequest}.
+   *
+   * @param request contains a set of input ids and a set of seed users.
+   */
+  private void collectRecommendations(SocialProofRequest request) {
+    socialProofs.clear();
+    socialProofWeights.clear();
+    LongSet inputIds = request.getRightNodeIds();
```
Comment (jerryjiangabc) [inline graphjet-core/src/main/java/com/twitter/graphjet/algorithms/socialproof/SocialProofGenerator.java]: inputIds => inputRightNodeIds

File: graphjet-core/src/test/java/com/twitter/graphjet/algorithms/salsa/SalsaBitmaskTest.java [hunk 13]
```diff
@@ -48,6 +48,7 @@
 
 public class SalsaBitmaskTest {
 
+  private final TweetIDMask tweetIDMask = new TweetIDMask();
```
Comment (jerryjiangabc) [inline graphjet-core/src/test/java/com/twitter/graphjet/algorithms/salsa/SalsaBitmaskTest.java]: consistent usage of variable modifiers.

File: graphjet-core/src/main/java/com/twitter/graphjet/algorithms/TweetIDMask.java [hunk 14]
```diff
@@ -20,11 +20,7 @@
 /**
  * The bit mask used to manipulate tweet ids to encode further information.
  */
-public final class TweetIDMask {
-
-  // Utility class
-  private TweetIDMask() {
-  }
+public final class TweetIDMask implements IDMask {
```
Comment (rossides) [inline graphjet-core/src/main/java/com/twitter/graphjet/algorithms/TweetIDMask.java]: Updated the other implementation to be final

File: graphjet-core/src/main/java/com/twitter/graphjet/algorithms/counting/tweet/TopSecondDegreeByCountTweetRecsGenerator.java [hunk 15]
```diff
@@ -36,6 +36,7 @@
 import it.unimi.dsi.fastutil.longs.LongSet;
 
 public final class TopSecondDegreeByCountTweetRecsGenerator {
+  private static TweetIDMask tweetIDMask = new TweetIDMask();
```
Comment (rossides) [inline graphjet-core/src/main/java/com/twitter/graphjet/algorithms/counting/tweet/TopSecondDegreeByCountTweetRecsGenerator.java]: Fixed

File: graphjet-core/src/main/java/com/twitter/graphjet/algorithms/salsa/SalsaSelectResults.java [hunk 16]
```diff
@@ -47,6 +47,7 @@
 
   private final CommonInternalState<T> salsaInternalState;
   private final SalsaStats salsaStats;
+  private final TweetIDMask tweetIDMask = new TweetIDMask();
```
Comment (rossides) [inline graphjet-core/src/main/java/com/twitter/graphjet/algorithms/salsa/SalsaSelectResults.java]: Updated everywhere to be private static final

File: graphjet-core/src/main/java/com/twitter/graphjet/algorithms/socialproof/SocialProofGenerator.java (hunk 11 above)
Comment (rossides) [inline graphjet-core/src/main/java/com/twitter/graphjet/algorithms/socialproof/SocialProofGenerator.java]: Fixed

File: graphjet-core/src/main/java/com/twitter/graphjet/algorithms/TweetIDMask.java (hunk 14 above)
Comment (guimingTang) [inline graphjet-core/src/main/java/com/twitter/graphjet/algorithms/TweetIDMask.java]: Curious about the benefit of declaring the masks final?

File: graphjet-core/src/test/java/com/twitter/graphjet/algorithms/socialproof/MomentSocialProofTest.java [hunk 17]
```diff
@@ -0,0 +1,92 @@
+/**
+ * Copyright 2016 Twitter. All rights reserved.
```
Comment (guimingTang) [inline graphjet-core/src/test/java/com/twitter/graphjet/algorithms/socialproof/MomentSocialProofTest.java]: nit: 2017

**Comment by guimingTang** [comment]
Could you also test to make sure current tweet & metadata recommendations work as before from the debug endpoints?

File: graphjet-core/src/main/java/com/twitter/graphjet/algorithms/TweetIDMask.java (hunk 14 above)
Comment (rossides) [inline graphjet-core/src/main/java/com/twitter/graphjet/algorithms/TweetIDMask.java]: Mostly maintaining the existing behavior, but the main thing about declaring final classes is to show that they should not be extended, which makes sense in this case. You should only extend IDMask, but not extend the subclasses in any way.

File: graphjet-core/src/test/java/com/twitter/graphjet/algorithms/socialproof/MomentSocialProofTest.java (hunk 17 above)
Comment (rossides) [inline graphjet-core/src/test/java/com/twitter/graphjet/algorithms/socialproof/MomentSocialProofTest.java]: Fixed

File: graphjet-core/src/main/java/com/twitter/graphjet/algorithms/salsa/SalsaSelectResults.java [hunk 18]
```diff
@@ -44,6 +44,7 @@
  */
 public class SalsaSelectResults<T extends LeftIndexedBipartiteGraph> {
   private static final Logger LOG = LoggerFactory.getLogger("graph");
+  private static final TweetIDMask TweetIDMask = new TweetIDMask();
```
Comment (jerryjiangabc) [inline graphjet-core/src/main/java/com/twitter/graphjet/algorithms/salsa/SalsaSelectResults.java]: this is java, so the static final field should be TWEET_ID_MASK. could you do a git grep and change it everywhere?

**Pull request merged**